### PR TITLE
Feature/1d mapper

### PR DESCRIPTION
**Pull request by damiansteiger**
Adds a LinearMapper

For more information and results, see our paper [arXiv:1806.01861](https://arxiv.org/abs/1806.01861)

### CODE REVIEW

File: projectq/cengines/_linearmapper.py [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+    """
+
+    def __init__(self, num_qubits, cyclic=False, storage=1000):
+        """
+        Initialize an linear chain mapper compiler engine.
+
+        Args:
+            num_qubits(int): Number of physical qubits in chain
+            cyclic(bool): If 1D chain is a cycle. Default is False.
+            storage: Number of gates to temporarily store
```
Comment (thomashaener) [inline projectq/cengines/_linearmapper.py]: storage (int): [...] default is 1000.

File: projectq/cengines/_linearmapper.py [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+        else:
+            return False
+
+    @staticmethod
+    def return_new_mapping(num_qubits, cyclic, currently_allocated_ids,
+                           stored_commands, current_mapping):
+        """
+        Builds a mapping of qubits to a linear chain.
+
+        It goes through self._saved_commands and tries to find a
```
Comment (thomashaener) [inline projectq/cengines/_linearmapper.py]: through stored_commands

File: projectq/cengines/_linearmapper.py [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+                for qubit in qureg:
+                    qubit_ids.append(qubit.id)
+
+            if len(qubit_ids) > 2 or len(qubit_ids) == 0:
+                raise Exception("Invalid command (number of qubits): " +
+                                str(cmd))
+
+            elif isinstance(cmd.gate, AllocateQubitGate):
+                qubit_id = cmd.qubits[0][0].id
+                if len(allocated_qubits) < num_qubits:
```
Comment (thomashaener) [inline projectq/cengines/_linearmapper.py]: I suggest `if qubit_id not in...`

**Comment by damiansteiger** [comment]
Thanks for your review of this long PR 👍 

Docs have been updated

**Pull request merged**